### DISCUSSION TITLE
[Detector Support]: Beelink EQ13 with Coral m.2 Dual Edge TPU

### DISCUSSION THREAD

**Discussion by templeowls21** [blakeblackshear/frigate#14343] [Ask A Question]
Hello. I am running the Frigate add-on in HA and its working great. I want to get detection up and running next.

Currently I have a Beelink EQ13 and I was planning to get a Coral m.2 Dual Edge TPU. The Beelink has a second m.2 slot.

Before I go an order the m.2 Coral instead of the USB version, I thought I'd check here to see if there's any problems I'd run into with this plan or if the USB version is the better route.

For what it's worth I have 10 IP cameras and I want to use detection on all of them (with a low-res sub stream for each)

**Comment by Strider0** [comment]
I'm running the m.2 Dual Edge in an EQ12 with 24 HD cameras doing detection on the sub streams.  It works well.  It initially ran hot so I adjusted the fan settings in the bios to be more aggressive and that brought the temps down.  My understanding is that you can only access one of the TPUs, but the performance has been great for me.

**Comment by davidsmoker** [comment]
Have done the same with frigate full acces. Pci and usb Coral . Works great

**Comment by rmunroe** [comment]
> Hello. I am running the Frigate add-on in HA and its working great. I want to get detection up and running next.
> 
> Currently I have a Beelink EQ13 and I was planning to get a Coral m.2 Dual Edge TPU. The Beelink has a second m.2 slot.
> 
> Before I go an order the m.2 Coral instead of the USB version, I thought I'd check here to see if there's any problems I'd run into with this plan or if the USB version is the better route.
> 
> For what it's worth I have 10 IP cameras and I want to use detection on all of them (with a low-res sub stream for each)

Did you ever get the Coral M.2 working in the Beelink EQ13? I've been wrestling with it all night and am starting to think I have a bad Coral.

**Reply by davidsmoker** [reply]
Yes working now for months

**Reply by davidsmoker** [reply]
Do you have disabled protection mode in frigate?

**Reply by rmunroe** [reply]
Thanks for the response. I couldn't even get that far. I was never able to get the OS to load the gasket driver and make `/dev/apex*` available. I tried using the official Google deb repo, and manually building. I even tried using fresh Ubuntu 22.04 and Debian 12 installs to no avail. I am returning the A+E key version and trying again with the B+M key version.

**Reply by davidsmoker** [reply]
I used to be in proxmox and EQ12, what a misery with those drivers for the coral. ( am not IT like many here ) With the EQ13 I did the ubuntu install from HA, removed wifi slot and replaced it with the Coral. 
Frigate full access installed and booted without protection mode, this worked immediately. Now have 7 cameras, detector inference speed 7.47ms.
I had read somewhere that the drivers are automatically loaded

**Reply by rmunroe** [reply]
Weird. Frigate docs [here](https://docs.frigate.video/frigate/hardware#google-coral-tpu) say that the Coral M.2 driver needs to be installed on the host. This is what I was unable to get past. I am running Ubuntu 24.04 on the EQ13 with only Frigate running in Docket, not as an add-on on my Home Assistant install, which is on another machine (via Proxmox). I'm still leaning towards having a bad Coral unit and will be returning it.

**Reply by ShawnCollinge** [reply]
Mines not working in my EQ13 either, I saw some other reports of it not working for others as well.

**Reply by NickM-27** [reply]
See https://deploy-preview-13787--frigate-docs.netlify.app/troubleshooting/edgetpu#pcie-coral-not-detected

**Reply by ShawnCollinge** [reply]
> See https://deploy-preview-13787--frigate-docs.netlify.app/troubleshooting/edgetpu#pcie-coral-not-detected

```
shawn@debian:~$ git clone https://github.com/jnicolson/gasket-builder
Cloning into 'gasket-builder'...
remote: Enumerating objects: 29, done.
remote: Counting objects: 100% (29/29), done.
remote: Compressing objects: 100% (20/20), done.
remote: Total 29 (delta 8), reused 28 (delta 7), pack-reused 0 (from 0)
Receiving objects: 100% (29/29), 4.48 KiB | 2.24 MiB/s, done.
Resolving deltas: 100% (8/8), done.
shawn@debian:~$ cd gasket-builder
shawn@debian:~/gasket-builder$ sudo docker build --output . .
[+] Building 55.4s (11/11) FINISHED                                                                                                                                    docker:default
 => [internal] load build definition from Dockerfile                                                                                                                             0.1s
 => => transferring dockerfile: 547B                                                                                                                                             0.0s
 => [internal] load metadata for docker.io/library/ubuntu:24.04                                                                                                                  1.2s
 => [internal] load .dockerignore                                                                                                                                                0.1s
 => => transferring context: 2B                                                                                                                                                  0.0s
 => [build-stage 1/6] FROM docker.io/library/ubuntu:24.04@sha256:80dd3c3b9c6cecb9f1667e9290b3bc61b78c2678c02cbdae5f0fea92cc6734ab                                                1.8s
 => => resolve docker.io/library/ubuntu:24.04@sha256:80dd3c3b9c6cecb9f1667e9290b3bc61b78c2678c02cbdae5f0fea92cc6734ab                                                            0.0s
 => => sha256:b1d9df8ab81559494794e522b380878cf9ba82d4c1fb67293bcf931c3aa69ae4 2.30kB / 2.30kB                                                                                   0.0s
 => => sha256:de44b265507ae44b212defcb50694d666f136b35c1090d9709068bc861bb2d64 29.75MB / 29.75MB                                                                                 0.9s
 => => sha256:80dd3c3b9c6cecb9f1667e9290b3bc61b78c2678c02cbdae5f0fea92cc6734ab 6.69kB / 6.69kB                                                                                   0.0s
 => => sha256:6e75a10070b0fcb0bead763c5118a369bc7cc30dfc1b0749c491bbb21f15c3c7 424B / 424B                                                                                       0.0s
 => => extracting sha256:de44b265507ae44b212defcb50694d666f136b35c1090d9709068bc861bb2d64                                                                                        0.7s
 => [build-stage 2/6] RUN apt-get update                                                                                                                                         6.0s
 => [build-stage 3/6] RUN apt-get install -q -y --no-install-recommends git curl devscripts dkms dh-dkms build-essential debhelper                                              42.1s
 => [build-stage 4/6] RUN git clone https://github.com/google/gasket-driver.git                                                                                                  0.8s
 => [build-stage 5/6] RUN cd gasket-driver && curl -L https://github.com/heitbaum/gasket-driver/commit/4b2a1464f3b619daaf0f6c664c954a42c4b7ce00.patch | git apply -v             0.5s
 => [build-stage 6/6] RUN cd gasket-driver && debuild -us -uc -tc -b                                                                                                             2.3s
 => [export-stage 1/1] COPY --from=build-stage *.deb .                                                                                                                           0.2s
 => exporting to client directory                                                                                                                                                0.0s
 => => copying files 49.07kB                                                                                                                                                     0.0s
shawn@debian:~/gasket-builder$ sudo dpkg -i gasket-dkms_1.0-18_all.deb
(Reading database ... 56954 files and directories currently installed.)
Preparing to unpack gasket-dkms_1.0-18_all.deb ...
Deprecated feature: REMAKE_INITRD (/var/lib/dkms/gasket/1.0/source/dkms.conf)
Deprecated feature: REMAKE_INITRD (/var/lib/dkms/gasket/1.0/source/dkms.conf)
Deprecated feature: REMAKE_INITRD (/var/lib/dkms/gasket/1.0/source/dkms.conf)
Module gasket-1.0 for kernel 6.1.0-29-amd64 (x86_64).
Before uninstall, this module version was ACTIVE on this kernel.

gasket.ko:
 - Uninstallation
   - Deleting from: /lib/modules/6.1.0-29-amd64/updates/dkms/
 - Original module
   - No original module was found for this module on this kernel.
   - Use the dkms install command to reinstall any previous module version.

apex.ko:
 - Uninstallation
   - Deleting from: /lib/modules/6.1.0-29-amd64/updates/dkms/
 - Original module
   - No original module was found for this module on this kernel.
   - Use the dkms install command to reinstall any previous module version.
depmod....
Deleting module gasket-1.0 completely from the DKMS tree.
Unpacking gasket-dkms (1.0-18) over (1.0-18) ...
Setting up gasket-dkms (1.0-18) ...
Loading new gasket-1.0 DKMS files...
Deprecated feature: REMAKE_INITRD (/usr/src/gasket-1.0/dkms.conf)
Building for 6.1.0-29-amd64
Building initial module for 6.1.0-29-amd64
Deprecated feature: REMAKE_INITRD (/var/lib/dkms/gasket/1.0/source/dkms.conf)
Done.
Deprecated feature: REMAKE_INITRD (/var/lib/dkms/gasket/1.0/source/dkms.conf)
Deprecated feature: REMAKE_INITRD (/var/lib/dkms/gasket/1.0/source/dkms.conf)

gasket.ko:
Running module version sanity check.
 - Original module
   - No original module exists within this kernel
 - Installation
   - Installing to /lib/modules/6.1.0-29-amd64/updates/dkms/

apex.ko:
Running module version sanity check.
 - Original module
   - No original module exists within this kernel
 - Installation
   - Installing to /lib/modules/6.1.0-29-amd64/updates/dkms/
depmod...
shawn@debian:~/gasket-builder$ lspci -nn | grep 089a
shawn@debian:~/gasket-builder$ sudo reboot

Broadcast message from root@debian on pts/1 (Mon 2025-01-13 09:10:35 PST):

The system will reboot now!

.......

shawn@debian:~$ lspci -nn | grep 089a
shawn@debian:~$ cd /dev/
shawn@debian:/dev$ ls
autofs           debian-vg  hpet       loop1         mei0     nvme0n1p1  ptmx      stderr  tty11  tty2   tty28  tty36  tty44  tty52  tty60  ttyS2        vcs3   vcsa5  vfio
block            disk       hugepages  loop2         mem      nvme0n1p2  pts       stdin   tty12  tty20  tty29  tty37  tty45  tty53  tty61  ttyS3        vcs4   vcsa6  vga_arbiter
btrfs-control    dm-0       hwrng      loop3         mqueue   nvme0n1p3  random    stdout  tty13  tty21  tty3   tty38  tty46  tty54  tty62  uhid         vcs5   vcsu   vhci
bus              dm-1       initctl    loop4         net      nvme1      rfkill    tpm0    tty14  tty22  tty30  tty39  tty47  tty55  tty63  uinput       vcs6   vcsu1  vhost-net
char             dri        input      loop5         ng0n1    nvme1n1    rtc       tpmrm0  tty15  tty23  tty31  tty4   tty48  tty56  tty7   urandom      vcsa   vcsu2  vhost-vsock
console          fd         kmsg       loop6         ng1n1    nvram      rtc0      tty     tty16  tty24  tty32  tty40  tty49  tty57  tty8   userfaultfd  vcsa1  vcsu3  watchdog
core             full       kvm        loop7         null     port       shm       tty0    tty17  tty25  tty33  tty41  tty5   tty58  tty9   vcs          vcsa2  vcsu4  watchdog0
cpu_dma_latency  fuse       log        loop-control  nvme0    ppp        snapshot  tty1    tty18  tty26  tty34  tty42  tty50  tty59  ttyS0  vcs1         vcsa3  vcsu5  zero
cuse             gpiochip0  loop0      mapper        nvme0n1  psaux      snd       tty10   tty19  tty27  tty35  tty43  tty51  tty6   ttyS1  vcs2         vcsa4  vcsu6
shawn@debian:/dev$
```

**Reply by ShawnCollinge** [reply]
Here's some posts that seem the EQ13+ doesn't allow the coral to be plugged into the wifi port:

https://www.reddit.com/r/BeelinkOfficial/comments/1hstjdn/eq14_m2_coral_accelerator_ae_key/

https://www.reddit.com/r/homeassistant/comments/1dxyqx0/comment/lg7bo2f/

**Reply by rmunroe** [reply]
Shawn's experience is exactly what I've tried as well (along with many other attempts of manually building, trying different OSes, etc.) It builds fine, but the device is never loaded.

**Reply by ShawnCollinge** [reply]
@rmunroe do you have the Dual edge or the A+E card? I'm using the A+E but it seems @davidsmoker is using the dual edge.

**Reply by rmunroe** [reply]
@ShawnCollinge non-dual, single A+E card.

![image](https://github.com/user-attachments/assets/da635ae4-4346-44d8-8b1f-97c30b4d24b9)

**Reply by ShawnCollinge** [reply]
Not sure why it would work differently but I just ordered the dual edge and will give it a go. I'll report back.

**Reply by rmunroe** [reply]
I bought the B+M key, which arrived today. I will let you know how that goes as well.

**Reply by rmunroe** [reply]
The B+M key worked with zero issue.

![IMG_2441](https://github.com/user-attachments/assets/dcb33d0b-be5e-4b7d-9466-319e066ff8ee)

I followed the instructions on the [Coral.ai website](https://coral.ai/docs/m2/get-started/#2a-on-linux), except for when it said to install gasket-dkms using apt, I built it with the [gasket-builder](https://github.com/jnicolson/gasket-builder) docker image and installed the output `.deb` with `dpkg -i gasket-dkms_1.0-18_all.deb`. Then I wrapped up the Coral.ai instructions, with the exception of not creating a group called `apex`, and changing the apex rule modification to use the existing `docker` group (which I read about while trying to figure out how to get the A+E card to work):

```bash
sudo sh -c "echo 'SUBSYSTEM==\"apex\", MODE=\"0660\", GROUP=\"docker\"' >> /etc/udev/rules.d/65-apex.rules"
```

After I rebooted, the coral board was immediately recognized and showed up via `lspci -nn | grep 089a` as they mention.

Adding the detector to my Frigate config was trivial, and now everything is working well. Averaging 6.6ms inference speed.

```yaml
detectors:
  coral:
    type: edgetpu
    device: pci
```

**Reply by ShawnCollinge** [reply]
@rmunroe Do you still have the wifi card that it came with? I think these boards use CNVio instead of pcie+usb for the wifi cards. It's the same slot but it doesn't work with pcie devices.
